File: src/primes.rb
class Primes

  def initialize
    @primes = []
  end

  def fill(np)
    @primes << 2
    n = 1
    p = 3
    begin
      while n<np
        isPrime = true
        j = 0
        last = @primes[0]
        while isPrime and last*last<p
          if p % last == 0
            isPrime = false
          else
            j += 1
            last = @primes[j]
          end
        end
        if isPrime
          @primes[n] = p
          n += 1
        end
        p += 2
      end
    rescue
      puts "Error on index %s" % n
    end
  end

  def getSize
    @primes.length
  end

  def getPrime(index)
    @primes[index]
  end

end

if ARGV.length<1
  puts "Argument: number of primes to compute"
else
  size = ARGV[0].to_i
  p = Primes.new
  t1 = Time.new.to_i
  p.fill(size)
  t2 = Time.new.to_i
  dt = t2-t1
  puts "Time spent: %s s" % dt
  puts "Number of elements: %s" % p.getSize
  puts "Last prime: %s" % p.getPrime(p.getSize-1)
end

File: src/primes.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define TRUE 1
#define FALSE 0
#define BOOLEAN int

long *primes;
int np;

void fill() {
	primes[0] = 2;
  int n = 1;
  long p = 3;
  while (n<np) {
    BOOLEAN isPrime = TRUE;
    int j = 0;
    long last = primes[0];
    while ((isPrime) && (last*last<p)) {
      if ((p % last)==0) isPrime = FALSE;
      else {
        last = primes[++j];
      }
    }
    if (isPrime) primes[n++] = p;
    p += 2;
  }
}

int main(int argc, char *argv[]) {
	if (argc<2) {
		return 1;
	}
	sscanf(argv[1], "%d", &np);
	primes = (long*) malloc(sizeof(long)*np);
	clock_t t1 = clock();
	fill();
	clock_t t2 = clock();
	long dt = (long) ((t2-t1)/1000);
	printf("CPU time spent: %ld ms\n", dt);
	printf("Number of elements: %d\n", np);
	printf("Last prime: %ld\n", primes[np-1]);
	free(primes);
	return 0;
}
